src/cnxml/jsx.ts

```diff
@@ -0,0 +1,472 @@
+// Copyright 2020 OpenStax Poland
+// Licensed under the MIT license. See LICENSE file in the project root for
+// full license text.
+
+import { Case } from '../interfaces'
+
+/** Qualified name of an element or attribute */
+export interface Name {
+    /**
+     * Namespace
+     *
+     * When omitted for an element name it is inherited form the parent element.
+     * When omitted for an attribute, the name refers to the default namespace.
+     */
+    namespace?: string
+    /** Local name */
+    local: string
+}
+
+/** An element */
+export interface Element {
+    /** Element's name */
+    name: Name
+    /** Attributes set on this element */
+    attributes: Omit<JSX.IntrinsicAttributes, 'children'> & { [key: string]: any }
+    /** Children nodes */
+    children: Node
+}
+
+/** Any value that can be used as child of a JSX element */
+export type Node = Element | string | Node[]
+
+/** JSX element creator */
+export function createElement<
+    K extends keyof JSX.IntrinsicElements,
+    A extends JSX.IntrinsicAttributes,
+>(
+    name: K,
+    attrs: A | null,
+    ...children: Node[]
+): Element {
+    const { xmlns, ...attributes } = attrs || {}
+
+    return {
+        name: { namespace: xmlns, local: name },
+        attributes,
+        children,
+    }
+}
+
+export const XML_NAMESPACE = 'http://www.w3.org/XML/1998/namespace'
+
+export const XMLNS_NAMESPACE = 'http://www.w3.org/2000/xmlns/'
+
+export const CNXML_NAMESPACE = 'http://cnx.rice.edu/cnxml'
+
+/** Natural language extensions to CNXML */
+export const CMLNLE_NAMESPACE = 'http://katalysteducation.org/cmlnle/1.0'
+
+/** CNXML extensions to facilitate better editing experience */
+export const EDITING_NAMESPACE = 'http://adaptarr.naukosfera.com/editing/1.0'
+
+/** Mapping from XML prefixes to namespace URIs */
+const NAMESPACE_PREFIXES: { [prefix: string]: string | undefined } = {
+    xml: XML_NAMESPACE,
+    cmlnle: CMLNLE_NAMESPACE,
+    editing: EDITING_NAMESPACE,
+}
+
+interface Renderer {
+    namespace: string
+    doc: Document
+    depth: number
+}
+
+/** Render a JSX element into an XML document */
+export function render(root: Element): Document {
+    const namespace = root.name.namespace
+    const doc = document.implementation.createDocument(
+        namespace, root.name.local, null)
+    const renderer = { namespace, doc, depth: 0 }
+
+    for (const [prefix, uri] of Object.entries(NAMESPACE_PREFIXES)) {
+        if (uri === XML_NAMESPACE) continue
+
+        doc.documentElement.setAttributeNS(XMLNS_NAMESPACE, `xmlns:${prefix}`, uri)
+    }
+
+    finishElement(renderer, root, doc.documentElement)
+
+    return doc
+}
+
+/** Render a JSX element into an XML element */
+function renderElement(renderer: Renderer, element: Element): globalThis.Element {
```

Comment:
Can this handle also `NodeListOf<ChildNode>`? In Adaptarr we are using SourceElement which is then taking all childNodes, parsing them as an XML and return their content